README.md
nyx
===

This code is very old and has not been maintained in a while. It is also not the best quality code and is most definitely riddled with race conditions and other nasty things. Use at own risk.


Changelog
I know this isn't correct ChangeLog format...
2009-11-16  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Fixed a UserTreeView.hide() bug in Windows(Weird..)

2009-11-15  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Nyx now checks the 4 value of a TreeIter to determine if it's a channel a server or a user
    * The server TreeIter now gets added in the connect function(IRC.py)
    * Added multi-server support! /newserver address port
    * Fixed a message buffer bug, where you can't send messages AT ALL when there is no channels joined...
    * Fixed a bug with cycle_address

2009-11-13  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Added more themes stuff
    * Fixed a Mode change bug
    * Cleaned up the code a bit.
    * Added /eval

2009-11-09  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Added some more themes stuff
    * Fixed some Encoding problems
    * Fixed problem with NOTICE AUTH, without ':irc.server.com' in front

2009-11-05  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Added themes, now the text that Nyx outputs will be customizable(Still not finished yet)

2009-10-31  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Nyx now reconnects to the server when disconnected
    * Fixed some mIRC color bugs

2009-10-23  Dominik Picheta  <dominikpicheta@googlemail.com>

    * added mIRC colors, bold and underline
    * Fixed the mode changes bug!

2009-10-03  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Made Private Messages work like they're suppose to...
    * Added disconnect messages...

2009-09-26  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Added settings, you can change them in settings.xml
    * Cleaned up the code A LOT, moved all the IRC Events to IRCEvents.py

2009-09-20  Dominik Picheta  <dominikpicheta@googlemail.com>

    * Changed the lag meter to show miliseconds, instead 0.123456789
    * Added channel mode info when you right click a channel
    * Fixed the problem with multiple users changed modes
    * Added SSL Support

2009-09-16  Dominik Picheta  <dominikpicheta@googlemail.com>

    * topicChange fix
    * InspIRCd server NICK fix

2009-09-09  Dominik Picheta  <dominikpicheta@googlemail.com>

    * KICK message, user list fix.

2009-08-31  Dominik Picheta  <dominikpicheta@googlemail.com>

    * UserList bug fixes
    * channel Topics are now shown
    * Nick change bug fix

2009-08-31  Dominik Picheta  <dominikpicheta@googlemail.com>

    * The UserList is now located on the right, instead of merged with the left TreeView.
      There is a peculiar GTKWarning which i don't know how to get rid of, it doesn't seem to couse any problems but might in some cases
      (It only shows the first time anything is added to the ListStore)
    * Added a way to close channels

2009-08-30  Dominik Picheta  <dominikpicheta@googlemail.com>

    * When joining a channel that channel gets automatically selected,
      and also it says "You have joined #channel"

2009-08-29  Dominik Picheta  <dominikpicheta@googlemail.com>


    * Fixed some user list problems:
    When a normal user joined, in some cases he appeared above a user with a mode.

    PongStuff.py(joinResp)[Line 379], The onUserJoin event(Which adds the user that joined
    to the TreeView) gets called instantaneously instead of using idle_add, this might couse a SEGFAULT!

    * Better debug information printing.
    

2009-08-28  Dominik Picheta  <dominikpicheta@googlemail.com>

    * MainForm.py (OnOwnPrivMsg): Line 1273,
    Added .lower()

    * MainForm.py (onPrivMsg): Line 485
    Added .lower()



mIRCColors.py
#!/usr/bin/env python
"""
Nyx - A powerful IRC Client
Copyright (C) 2009 Mad Dog Software 
http://maddogsoftware.co.uk - morfeusz8@yahoo.co.uk

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
"""
import gtk
import pygtk

mIRCColors=dict()
mIRCColors[0]=gtk.gdk.Color(red=204 * 257, green=204 * 257, blue=204 * 257, pixel=0)
mIRCColors[1]=gtk.gdk.Color(red=0, green=0, blue=0, pixel=0)
mIRCColors[2]=gtk.gdk.Color(red=54 * 257, green=54 * 257, blue=178 * 257, pixel=0)
mIRCColors[3]=gtk.gdk.Color(red=42 * 257, green=140 * 257, blue=42 * 257, pixel=0)
mIRCColors[4]=gtk.gdk.Color(red=195 * 257, green=59 * 257, blue=59 * 257, pixel=0)
mIRCColors[5]=gtk.gdk.Color(red=199 * 257, green=50 * 257, blue=50 * 257, pixel=0)
mIRCColors[6]=gtk.gdk.Color(red=128 * 257, green=38 * 257, blue=127 * 257, pixel=0)
mIRCColors[7]=gtk.gdk.Color(red=102 * 257, green=54 * 257, blue=31 * 257, pixel=0)
mIRCColors[8]=gtk.gdk.Color(red=217 * 257, green=166 * 257, blue=65 * 257, pixel=0)
mIRCColors[9]=gtk.gdk.Color(red=61 * 257, green=204 * 257, blue=61 * 257, pixel=0)
mIRCColors[10]=gtk.gdk.Color(red=26 * 257, green=85 * 257, blue=85 * 257, pixel=0)
mIRCColors[11]=gtk.gdk.Color(red=47 * 257, green=140 * 257, blue=116 * 257, pixel=0)
mIRCColors[12]=gtk.gdk.Color(red=69 * 257, green=69 * 257, blue=230 * 257, pixel=0)
mIRCColors[13]=gtk.gdk.Color(red=176 * 257, green=55 * 257, blue=176 * 257, pixel=0)
mIRCColors[14]=gtk.gdk.Color(red=76 * 257, green=76 * 257, blue=76 * 257, pixel=0)
mIRCColors[15]=gtk.gdk.Color(red=149 * 257, green=149 * 257, blue=149 * 257, pixel=0)

mIRCColors[16]=gtk.gdk.Color(red=204 * 257, green=204 * 257, blue=204 * 257, pixel=0)
mIRCColors[17]=gtk.gdk.Color(red=0, green=0, blue=0, pixel=0)
mIRCColors[18]=gtk.gdk.Color(red=54 * 257, green=54 * 257, blue=178 * 257, pixel=0)
mIRCColors[19]=gtk.gdk.Color(red=42 * 257, green=140 * 257, blue=42 * 257, pixel=0)
mIRCColors[20]=gtk.gdk.Color(red=195 * 257, green=59 * 257, blue=59 * 257, pixel=0)
mIRCColors[21]=gtk.gdk.Color(red=255 * 257, green=0 * 257, blue=0 * 257, pixel=0)#Changed this because it's dull...
mIRCColors[22]=gtk.gdk.Color(red=128 * 257, green=38 * 257, blue=127 * 257, pixel=0)
mIRCColors[23]=gtk.gdk.Color(red=102 * 257, green=54 * 257, blue=31 * 257, pixel=0)
mIRCColors[24]=gtk.gdk.Color(red=217 * 257, green=166 * 257, blue=65 * 257, pixel=0)
mIRCColors[25]=gtk.gdk.Color(red=61 * 257, green=204 * 257, blue=61 * 257, pixel=0)
mIRCColors[26]=gtk.gdk.Color(red=26 * 257, green=85 * 257, blue=85 * 257, pixel=0)
mIRCColors[27]=gtk.gdk.Color(red=47 * 257, green=140 * 257, blue=116 * 257, pixel=0)
mIRCColors[28]=gtk.gdk.Color(red=50 * 257, green=150 * 257, blue=255 * 257, pixel=0)#Changed this because it's dull...
mIRCColors[29]=gtk.gdk.Color(red=176 * 257, green=55 * 257, blue=176 * 257, pixel=0)
mIRCColors[30]=gtk.gdk.Color(red=76 * 257, green=76 * 257, blue=76 * 257, pixel=0)
mIRCColors[31]=gtk.gdk.Color(red=149 * 257, green=149 * 257, blue=149 * 257, pixel=0)


def canonicalColor(s, bg=False, shift=0):
    """Assigns an (fg, bg) canonical color pair to a string based on its hash
    value.  This means it might change between Python versions.  This pair can
    be used as a *parameter to mircColor.  The shift parameter is how much to
    right-shift the hash value initially.
    """
    h = hash(s) >> shift
    fg = h % 14 + 2 # The + 2 is to rule out black and white.
    if bg:
        bg = (h >> 4) & 3 # The 5th, 6th, and 7th least significant bits.
        if fg < 8:
            bg += 8
        else:
            bg += 2
        return (fg, bg)
    else:
        return (fg, None)
